### PR TITLE
Scope system-prompt identifier substitution to live interpolations

### LINKED ISSUE #930: `--apply` rewrites prompt prose that names its own interpolation placeholder: `$CLAUDE_JOB_DIR/tmp` ships as `$e/tmp`
On 2.1.221, `--apply` rewrites prompt prose that names its own interpolation placeholder.

```diff
-Use `$CLAUDE_JOB_DIR/tmp` (`${U2p.join(e,"tmp")}`) for any temporary files
+Use `$e/tmp` (`${U2p.join(e,"tmp")}`) for any temporary files
```

The interpolation's fine. The prose isn't. `$CLAUDE_JOB_DIR` is an env var the prompt documents, and it's shipping to Claude Code as `$e`.

## Where it happens

`applyIdentifierMapping` in `src/systemPromptSync.ts`, line 1466 on `main`:

```js
for (const [humanName, actualVar] of sortedEntries) {
  const pattern = new RegExp(`\\b${humanName}\\b`, 'g');
  // Use a replacer function to avoid special replacement pattern interpretation (e.g., $$ -> $), see #237
  result = result.replace(pattern, () => actualVar);
}
```

That runs over the whole prompt body. Nothing requires the match to be inside a `${...}`. `CLAUDE_JOB_DIR` is this prompt's placeholder for minified `e`, so every occurrence becomes `e`, prose included.

## Sync isn't the culprit

The generated markdown still holds the right text:

```
$ grep -c '\$e/tmp' system-prompt-background-session-instructions.md
0
$ grep -o 'Use.\{0,66\}' system-prompt-background-session-instructions.md | head -1
Use \`$CLAUDE_JOB_DIR/tmp\` (\`${PATH_MODULE.join(CLAUDE_JOB_DIR,"tmp
```

So it's introduced on apply.

## Don't fix it by skipping the name

That last line uses the name both ways. Inside `${...}` it has to be substituted. In the prose it must not be.

Skip it outright and `PATH_MODULE.join(CLAUDE_JOB_DIR,"tmp")` goes into the bundle referencing something that doesn't exist there, which is #900.

## Scope

One prompt, `system-prompt-background-session-instructions`, but not one version. Sweeping every snapshot in `data/prompts`, the same corrupted prose appears in 87 of them, reaching back to `prompts-2.1.119.json`. The snapshot data is correct throughout. Only the write path is wrong.

Decoded prompt length on 2.1.221 goes 863 to 850.

Blast radius is one prompt today, but the mechanism is general. Placeholder names get picked to be human-meaningful (`CLAUDE_JOB_DIR`, `PATH_MODULE`, `TOOLS`), which is exactly what makes them likely to turn up in prompt prose.

## Nothing currently catches it

The output is a valid string literal in valid JavaScript. `node --check` passes, the binary boots, `claude -p` finishes a turn. The #908 parse gate can't see it because nothing's malformed.

It only shows up in a round-trip byte-identity comparison across the corpus, and only if you decode both sides first. Most apply-time rewrites are harmless escape respellings (`\xA7` for `§`, hex case), so an undecoded diff buries it. On 2.1.221: 661 sites, 594 byte-identical, 50 rewritten but decode-equal, 17 with genuinely different text. Sixteen of those are #922. This is the seventeenth.

## Fix direction

Scope the substitution to interpolation regions.

`interpolationRegions` is already in the same file from #902, and its doc comment states the rule the substitution side is missing:

> Identifier names are matched with word boundaries against the interpolation text only (not the whole body), so a name that also appears in ALL-CAPS prose (e.g. a "## TOOLS" heading) does not mask a drifted `${TOOLS}`

The detection side learned this one. The substitution side didn't.

One caveat for whoever picks it up: `interpolationRegions` returns region text, not offsets, and a replacement needs spans. Better to have the walker return spans and let the existing caller map those to text, so there's still exactly one parser. #922 has three attempts at a second, divergent copy of that walker, and all three desynchronised.

### PR DESCRIPTION
**Pull request by StreamDemon**
Fixes #930.

`applyIdentifierMapping` substituted each placeholder name across the whole prompt body, so prose that happens to name one of the prompt's own placeholders got rewritten to a minified variable.

```diff
-Use `$CLAUDE_JOB_DIR/tmp` (`${U2p.join(e,"tmp")}`) for any temporary files
+Use `$e/tmp` (`${U2p.join(e,"tmp")}`) for any temporary files
```

`system-prompt-background-session-instructions` documents the `$CLAUDE_JOB_DIR` environment variable and also uses `CLAUDE_JOB_DIR` as its placeholder for minified `e`. Both sit in the same sentence, so skipping the name isn't an option. The occurrence inside `${PATH_MODULE.join(CLAUDE_JOB_DIR,"tmp")}` still has to be mapped, and leaving that unmapped is #900.

This is older than it looks. The same prose has been shipping corrupted since `prompts-2.1.119.json`, across 87 prompt versions.

## What changed

The substitution is scoped by position instead of by name. Only text inside an unescaped `${...}` gets rewritten.

`interpolationRegions`, added for #902, is refactored into `interpolationSpans` with `interpolationRegions` as a thin wrapper over it, so the drift check and the substitution share one walker. #922 records three attempts at a second, divergent copy of that walker and all three desynced, so this avoids a fourth.

A span ends at `j - 1` only when the walker consumed a closing brace. When it stops because the content ran out, the interpolation is unterminated and the last character is prompt text rather than a delimiter. Ending at `j - 1` there drops that character and exposes a word boundary that doesn't exist, so `${FOOX` would substitute `FOO`. Ending at `j` keeps the text intact. Reconstructing all 61717 prompt bodies across every shipped snapshot, this changes `interpolationRegions` output on none of them, so drift detection is unaffected.

The per-name patterns are compiled once rather than rebuilt for every span.

## Verification

The question that decides whether scoping is safe is whether any identifier insertion point can fall outside a span, because that would leave a placeholder name in `cli.js` as an undefined variable. Across every snapshot in `data/prompts`:

| | |
| --- | --- |
| snapshots | 247 |
| prompts carrying identifiers | 24863 |
| identifier insertion points | 105218 |
| insertion points outside a span | **0** |
| prompt versions whose output changes | 87 |
| prompts losing a substitution | **0** |

All 87 changed versions are the same background-session prose.

That comparison runs against `main`'s own `applyIdentifierMapping`, not a reimplementation of it. Running the same harness with upstream's function on both sides reports 0 changed, which is what makes the 87 mean anything. A 400000-case fuzz over an alphabet of braces, quotes, backslashes, astral characters and the placeholder names finds no input where the fix adds a substitution upstream did not make.

Real 2.1.221 native install, isolated `TWEAKCC_CONFIG_DIR` and `TWEAKCC_CC_INSTALLATION_PATH` pointed at a copy of a pristine binary:

- round-trip over 661 prompt sites: byte-identical 594 to 595, real content changes 17 to 16
- the 16 that remain are the #922 escape-doubling set, unchanged
- the patched bundle carries `$CLAUDE_JOB_DIR/tmp` once and `$e/tmp` zero times
- 0 `✗`, 0 "Could not find", `node --check` passes, boots `2.1.221 (Claude Code)` + `4.3.2 (tweakcc)`, and `claude -p` completes a turn

492 tests, lint clean. Every guard the walker depends on is mutation-covered, and reverting any one of them turns exactly one test red: the escaped `\${` check, the in-string tracking, nested-brace depth, the cursor advance past a completed span, the global flag on the per-name patterns, and the unterminated-span end.

## Known limitation

The walker treats `'` as a string opener, so an apostrophe inside a nested template literal (`What's working`, `the user's request`) opens a quote it never sees closed and the span runs to the end of the content. Substitution stays effectively global for that tail, so #930 is not fixed for those prompts.

That behaviour is unchanged from today, where substitution is global everywhere, so nothing regresses. Making the walker understand nested template literals well enough to close the gap is the problem #922 documents three failed attempts at, and it belongs with that issue rather than here.

## Summary by CodeRabbit

* **Improvements**
  * The `--apply` option now substitutes interpolation placeholder names only within live template expressions (`${...}`), preserving matching text in regular prose.
  * Processing correctly handles nested expressions, escaped interpolations, quoted strings, adjacent expressions, and incomplete expressions.

* **Tests**
  * Added comprehensive coverage for interpolation substitution, including nested structures, edge cases, escaped content, and literal braces.

* **Documentation**
  * Added an Unreleased changelog entry describing the updated substitution behavior.

### CODE REVIEW

File: src/systemPromptSync.ts [hunk 1]
```diff
@@ -1457,17 +1482,42 @@ const applyIdentifierMapping = (
...
+  let result = '';
+  let cursor = 0;
+
+  for (const { start, end } of interpolationSpans(content)) {
+    let region = content.slice(start, end);
+
+    for (const [pattern, actualVar] of patterns) {
+      // Use a replacer function to avoid special replacement pattern interpretation (e.g., $$ -> $), see #237
+      region = region.replace(pattern, () => actualVar);
+    }
```
[the remark this answers is not in the document]
Comment (StreamDemon) [reply src/systemPromptSync.ts:1521]: Checked all three parts against the corpus. They're real properties of `\b`-based matching, but none of them is introduced here and none is reachable today.

Your example produces byte-identical output before and after this PR:

```
in       : ${PATH_MODULE.join("CLAUDE_JOB_DIR", CLAUDE_JOB_DIR)}
upstream : ${U2p.join("e", e)}
this PR  : ${U2p.join("e", e)}
identical: true
```

`main` replaces across the whole body, so it already rewrites that quoted argument. This PR only narrows where replacement happens, so its output is a strict subset of what ships today. The case it does change is the one the issue is about:

```
in       : Use `$CLAUDE_JOB_DIR/tmp` (`${PATH_MODULE.join(CLAUDE_JOB_DIR,"tmp")}`)
upstream : Use `$e/tmp` (`${U2p.join(e,"tmp")}`)
this PR  : Use `$CLAUDE_JOB_DIR/tmp` (`${U2p.join(e,"tmp")}`)
```

Reachability across every snapshot in `data/prompts`, over the 24626 prompt bodies that carry identifiers:

| case | occurrences |
| --- | --- |
| name inside an inert quoted string within an interpolation | 0 |
| name preceded by `$`, the `$FOO` case | 0 |
| `identifierMap` name that isn't `/^\w+$/` | 0 |

On the third one, `\b$FOO\b` compiles rather than throwing and then silently never matches, so a `$`-leading name would be a quiet no-op rather than an error. There aren't any: 742 distinct names across all snapshots, shortest 4 characters, 733 of them all-caps.

Worth flagging a correction to my own first measurement, because it's the interesting part. Counting names inside any nested literal gives 9965 hits, but nearly all of them sit inside nested *template* literals, like ``${X?`... ${TOOL_NAME.name} ...`:""}``. Those are live interpolations that have to be substituted. Only quote-delimited strings are inert here, and there are none.

On the proposed remedy: scanning token locations, and using them for drift detection too, means teaching the walker to parse nested template literals. That's the problem #922 records three failed attempts at, and it would change `hasIdentifierDrift` behaviour, which this PR deliberately leaves byte-identical across all 61717 prompt bodies. It's a legitimate follow-up, just a much larger one than a fix whose blast radius is one prompt.

One change pushed in response: a comment recording that "every identifier slot lands inside an interpolation" is an empirical property of the extractor's output rather than something the code enforces, along with the measurement behind it and what to re-check if that output shape ever changes.
